### PR TITLE
Org parser

### PR DESCRIPTION
**Pull request by Pierre-Andre**
Org: proposition of a new optlib parser for org-mode file

### CODE REVIEW

File: optlib/org.ctags [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
+# USA.
+#
+#  References:
+#  - https://orgmode.org/
+#
+--langdef=org
```
Comment (masatake) [inline optlib/org.ctags:25]: Could you use "Org" instead of "org"?

File: optlib/org.ctags [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# definition of levels
+--_tabledef-org=toplevel
+--_tabledef-org=srcblock
+################################
+# beginning of toplevel
+################################
+# if encounter a src block do language identify languge and defer
+--_mtable-regex-org=toplevel/^[:blank:]*#\+begin_src[ ]+([a-zA-Z0-9][-#+a-zA-Z0-9]*)//{tenter=srcblock}{_guest=\1,0end,}
+# sections stuff
+--_mtable-regex-org=toplevel/^\*\*\*\*\*\*\*[ \t]+([[:graph:][:blank:]]+)([\n])?/\1/G,subparagraph/
```
Comment (masatake) [inline optlib/org.ctags:45]: "G" kind is defined with `--kindef-...` option. So you have to specify only "G"; ",subparagraph" is redundant.

File: win32/ctags_vs2013.vcxproj [hunk 3]
```diff
@@ -460,4 +461,4 @@
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets">
   </ImportGroup>
-</Project>
\ No newline at end of file
+</Project>
```
Comment (masatake) [inline win32/ctags_vs2013.vcxproj:464]: This change should not be included.

File: Units/parser-org.r/simple-org.d/args.ctags [hunk 4]
```diff
@@ -0,0 +1 @@
+--extras=+g
```
Comment (masatake) [inline Units/parser-org.r/simple-org.d/args.ctags:1]: For extending the test case in the future, could you add `--sort=no` here?

**Comment by masatake** [comment]
Thank you for your contribution.
I made some minor comments.

File: optlib/org.ctags (hunk 1 above)
Comment (Pierre-Andre) [reply optlib/org.ctags:25]: I am not a github user and I do not know what I am supposed to do now:
1) do all your suggested changes and make another pull (using git pull only or the gui tool from github)
2) doing nothing and you take care of all changes on your side

Thank you for all your wonderful work on unversal-ctags !!!

File: optlib/org.ctags (hunk 2 above)
Comment (Pierre-Andre) [reply optlib/org.ctags:45]: Ok, same as above

File: Units/parser-org.r/simple-org.d/args.ctags (hunk 4 above)
Comment (Pierre-Andre) [reply Units/parser-org.r/simple-org.d/args.ctags:1]: Yes indeed, but same comment as above, I don't know how to do it in the "github collaborative project" way

File: optlib/org.ctags (hunk 1 above)
Comment (masatake) [reply optlib/org.ctags:25]: > I am not a github user and I do not know what I am supposed to do now:

If you are not a github user, I wonder how you can make this mostly perfect pull request :-)

3. do all my suggested changes then do 'git push --force ...' to overwrite the changes in this pull request.

Comment (masatake) [reply optlib/org.ctags:25]: If you don't know `git push --force ` well, you can make a new pull request instead.

**Comment by Pierre-Andre** [comment]
Dear Dr Yamato

I have updated using git push and I hope that all your suggestions will be taken into account in this magical git push (I do not understand where git know that my push will be added to that pull request but I supposed that it is done automagically in github when opening a PR).

Still I have a question for an improvement: some languages have aliases and org mode uses some:
sh/shell ; elisp/emacs-lisp... Is there a simple way in optlib parser to adjust these aliases before to use a guest parser
as the _guest flag needs the language name and not its alias.

Thanks for all

**Comment by masatake** [comment]
Thank you for updating. I will review it.
The setting for continuous development environments looks broken. I'm fixing them now.

> I have updated using git push and I hope that all your suggestions will be taken into account in this magical git push (I do not understand where git know that my push will be added to that pull request but I supposed that it is done automagically in github when opening a PR).

I expect you will contribute to this project more in the future:-)
I highly recommend you to study the following git commands:

* git rebase -i master
  - fixup
  - reword
  - pick
* git push --force
* git commit --fixup=....

With the three commands, you can rearrange your pull request as you want.

> Still I have a question for an improvement: some languages have aliases and org mode uses some:
sh/shell ; elisp/emacs-lisp... Is there a simple way in optlib parser to adjust these aliases before to use a guest parser
as the _guest flag needs the language name and not its alias.

Do you mean you want to run a guest parser specifying its alias from an optlib-based host parser?
If not, I need more examples of your question. Feel free to make an issue for the question.
I will focus on merging the Org parser here.

File: optlib/org.ctags (hunk 1 above)
Comment (masatake) [reply optlib/org.ctags:25]: This is not fixed yet.

File: win32/ctags_vs2013.vcxproj (hunk 3 above)
Comment (masatake) [reply win32/ctags_vs2013.vcxproj:463]: This is not fixed yet.

File: win32/ctags_vs2013.vcxproj.filters [hunk 5]
```diff
@@ -884,4 +887,4 @@
       <Filter>Resource Files</Filter>
     </ResourceCompile>
   </ItemGroup>
-</Project>
\ No newline at end of file
+</Project>
```
Comment (masatake) [inline win32/ctags_vs2013.vcxproj.filters:890]: This change is not needed; this is nothing to do with Org.

**Review by masatake — CHANGES_REQUESTED** [review]

File: optlib/org.ctags [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
+# USA.
+#
+#  References:
+#  - https://orgmode.org/
+#
+--langdef=Org
+--map-org=+.org
+# definitions
+--kinddef-org=d,definition,definitions
```
Comment (masatake) [inline optlib/org.ctags:28]: All org must be replaced with Org.

File: optlib/org.ctags [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
+# USA.
+#
+#  References:
+#  - https://orgmode.org/
+#
+--langdef=Org
+--map-org=+.org
```
Comment (masatake) [inline optlib/org.ctags:26]: This should be `--map-Org=+.org`.

File: win32/ctags_vs2013.vcxproj (hunk 3 above)
Comment (masatake) [reply win32/ctags_vs2013.vcxproj:463]: Again. No newline is needed here, at the end of file.

File: win32/ctags_vs2013.vcxproj.filters (hunk 5 above)
Comment (masatake) [reply win32/ctags_vs2013.vcxproj.filters:887]: Don't put the new line at the end of file.

File: win32/ctags_vs2013.vcxproj.filters [hunk 8]
```diff
@@ -249,9 +249,6 @@
     <ClCompile Include="..\optlib\mesonOptions.c">
       <Filter>Source Files\optlib</Filter>
     </ClCompile>
-    <ClCompile Include="..\optlib\org.c">
-      <Filter>Source Files\optlib</Filter>
-    </ClCompile>
```
Comment (masatake) [inline win32/ctags_vs2013.vcxproj.filters]: You should not remove these lines.

**Comment by masatake** [comment]
I will take over this pull request.

**Comment by masatake** [comment]
I will merge your parser via #3294.

**Comment by masatake** [comment]
Merged via #3294.

**Comment by masatake** [comment]
Thank you for your contribution.

**Pull request closed without merging**